### PR TITLE
feat: allow other peers to query blocks by range

### LINKED ISSUE #275: [P2P] Allow other peers to query blocks by range
We need to implement this handler, in which peers ask for a number of blocks by range. We should fetch these blocks from storage:

https://github.com/lambdaclass/lambda_ethereum_consensus/blob/1aeb3477e9eb940671b1fcc7ecf652d654c9e34f/lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex#L124

Here is the corresponding spec: https://github.com/lambdaclass/lambda_ethereum_consensus/blob/main/docs/specs/p2p-interface.md#beaconblocksbyrange-v2

Additionally, you should be familiarised with the req/resp specification: https://github.com/lambdaclass/lambda_ethereum_consensus/blob/main/docs/specs/p2p-interface.md#reqresp-interaction

### PR DESCRIPTION
**Pull request by Godspower-Eze**
**Description**

Allow peers request beacon blocks by range

Closes #275

### CODE REVIEW

File: lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex
Comment (MegaRedHand) [inline lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex:153]: This wouldn't work, sadly. The spec shows the response as an SSZ container, but that isn't actually correct. The response is given as a stream of chunks, each corresponding to a block, and prefixed with a header. An example would be:

```
request: 3 blocks: A, B (we don't have B)
response: `0 result (OK)` <> `ForkDigest-context` <> `size varint` <> `A (ssz_snappy encoded)` <>
           `3 result (not found)` <> `error_message`
```
Some relevant links:
- https://github.com/lambdaclass/lambda_ethereum_consensus/blob/d3344795df9d55c52c84da84ef12da8695c57243/docs/specs/p2p-interface.md?plain=1#L770-L771
- https://github.com/lambdaclass/lambda_ethereum_consensus/blob/main/docs/specs/p2p-interface.md?plain=1#L695

**Review by MegaRedHand — CHANGES_REQUESTED** [review]
Left some comments.

**Comment by MegaRedHand** [comment]
Leaving some context here (@Godspower-Eze please correct me if I'm wrong):

Networking configuration wasn't added to the config, but because `MAX_REQUEST_BLOCKS` will be modified (now `MAX_REQUEST_BLOCKS_DENEB`, all of them will be added to the config starting with Deneb. Because of that, we should start adding them manually to the config.

File: config/configs/mainnet.yaml
Comment (Godspower-Eze) [reply config/configs/mainnet.yaml:92]: That's the one thing I changed.

File: config/configs/minimal.yaml
Comment (Godspower-Eze) [reply config/configs/minimal.yaml:92]: That's the only thing I changed here too

File: lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex
Comment (Godspower-Eze) [reply lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex:145]: This returns the latest block received in a slot.

File: config/configs/mainnet.yaml
Comment (MegaRedHand) [reply config/configs/mainnet.yaml:92]: I was talking about the whitespace changes (like [here](https://github.com/lambdaclass/lambda_ethereum_consensus/pull/363/files#diff-1ff307241ed7036b050f1309475f079bf3a87977db78508c5701fd1bd23b8972L54-L56), [here](https://github.com/lambdaclass/lambda_ethereum_consensus/pull/363/files?diff=unified&w=0#diff-1ff307241ed7036b050f1309475f079bf3a87977db78508c5701fd1bd23b8972L43-R40), and [here](https://github.com/lambdaclass/lambda_ethereum_consensus/pull/363/files?diff=unified&w=0#diff-1ff307241ed7036b050f1309475f079bf3a87977db78508c5701fd1bd23b8972L21-L22)). They were probably done by a formatter. If you are using VSCode, you should use [the "save without formatting" command](https://stackoverflow.com/a/63480681) for this

File: lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex
Comment (MegaRedHand) [reply lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex:145]: That's how it works now, but in the future, it will return the block corresponding to that slot.

Comment (Godspower-Eze) [reply lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex:145]: Yeah, I was a bit unclear how many blocks should be in a slot. Is this defined?

**Comment by Godspower-Eze** [comment]
> Leaving some context here (@Godspower-Eze please correct me if I'm wrong):
> 
> Networking configuration wasn't added to the config, but because `MAX_REQUEST_BLOCKS` will be modified (now `MAX_REQUEST_BLOCKS_DENEB`, all of them will be added to the config starting with Deneb. Because of that, we should start adding them manually to the config.

Can you please rephrase? I am struggling to understand.

File: lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex
Comment (Godspower-Eze) [reply lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex:153]: result in this case is like a status_code, right?

File: lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex
Comment (MegaRedHand) [reply lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex:145]: 0~1 blocks per slot, as per the spec. While fork choice is validating the blocks there might be many, but this is quickly resolved, since:
- either they are blocks proposed by a non-proposer (which are quickly discarded),
- or they are from the proposer and the network should choose only one on the next block (and the proposer will be slashed)

**Comment by MegaRedHand** [comment]
Before, the config files didn't have the networking constants (I guess because they forgot to add them). Since the next fork ("Deneb") modifies some of those, they will be adding the networking constants to the YAMLs. Because of this, and because of those constants being labeled as _configuration_ in the spec, it's safe to have them in the YAMLs.

File: lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex
Comment (Godspower-Eze) [reply lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex:145]: Thank you. Please send a link where I can read up.

Comment (MegaRedHand) [reply lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex:145]: The information is in https://eth2book.info/capella/part2/consensus/, but I don't know if there's a single page that explains this.

File: config/configs/mainnet.yaml
Comment (mpaulucci) [reply config/configs/mainnet.yaml:92]: We should maybe configure a yaml formatter so that this doesn't happen?

File: lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex
Comment (mpaulucci) [inline lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex:123]: what does the 24 reference to?

Comment (Godspower-Eze) [reply lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex:123]: That's the length of the `BeaconBlocksByRangeRequest` SSZ Container.

File: lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex
Comment (MegaRedHand) [reply lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex]: @Godspower-Eze I just found out the `BeaconBlocksByRange` request doesn't send N chunks for N blocks:

https://github.com/lambdaclass/lambda_ethereum_consensus/blob/15617fd64f66daacf52b2e1545ad2a6efb81dd27/docs/specs/p2p-interface.md?plain=1#L898-L899

In this case, we should find a way to represent the "no blocks for this slot" in our DB, and parse that result accordingly here (not returning blocks for empty slots, instead of returning a response code of 3).

Comment (Godspower-Eze) [reply lib/lambda_ethereum_consensus/p2p/incoming_request_handler.ex]: Gotcha

File: lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex
Comment (MegaRedHand) [inline lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex:121]: You can use `min` for this

File: lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex
Comment (MegaRedHand) [inline lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex:127]: Nit
```suggestion
        |> Enum.map(&BlockStore.get_block_by_slot/1)
```

File: lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex
Comment (MegaRedHand) [inline lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex:131]: You can use `Enum.map` and then `Enum.join` to simplify this. Also, this:

```elixir
fn a ->
  case a do
    {:ok, ...} -> ...
    {:error, ...} -> ...
  end
end
```

is equivalent to this:

```elixir
fn
  {:ok, ...} -> ...
  {:error, ...} -> ...
end
```

File: lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex
Comment (MegaRedHand) [inline lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex:15]: typo

```suggestion
  @error_message_resource_available <<?R, ?e, ?s, ?o, ?u, ?r, ?c, ?e, 32, ?U, ?n, ?a, ?v, ?a, ?i,
```

File: lib/lambda_ethereum_consensus/store/block_store.ex
Comment (MegaRedHand) [inline lib/lambda_ethereum_consensus/store/block_store.ex:38]: Is this dead code?

File: lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex
Comment (MegaRedHand) [inline lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex:148]: Error messages are also `ssz_snappy`-encoded

**Review by MegaRedHand — COMMENTED** [review]
LGTM, but error messages must be SSZ serialized and Snappy compressed according to the spec

File: lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex
Comment (mpaulucci) [inline lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex:15]: nit: `error_message_resource_unavailable`

File: lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex
Comment (mpaulucci) [inline lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex:159]: maybe be more specific here, you need to match `:not_found`

**Review by MegaRedHand — COMMENTED** [review]
After this comment, I think we're good to merge

File: lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex
Comment (MegaRedHand) [reply lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex:15]: Also, the comment mentions "Resource Available" when it should be "Resource Unavailable"

**Review by MegaRedHand — APPROVED** [review]
The things that are left aren't really blocking (see `resource_available` typo), so I'll just leave my approve here and defer to @mpaulucci

**Review by mpaulucci — APPROVED** [review]

**Pull request merged**
